src/Uno.Toolkit.UI/Controls/ZoomContentControl/ZoomContentControl.xaml.cs

```diff
@@ -0,0 +1,777 @@
+
+		return rules;
+	}
+
+	private bool CanScrollUp() => !ContentBoundsVisibility.TopEdge;
+
+	private bool CanScrollDown() => !ContentBoundsVisibility.BottomEdge;
+
+	private bool CanScrollLeft() => !ContentBoundsVisibility.LeftEdge;
+
+	private bool CanScrollRight() => !ContentBoundsVisibility.RightEdge;
+
+	//Slide move is always on the opposite direction of the drag
+	private async void UpdateVerticalScrollBarValue()
+	{
+		VerticalScrollValue = VerticalOffset;
+		await RaiseRenderedContentUpdated();
+	}
+
+	//Slide move is always on the opposite direction of the drag
+	private async void UpdateHorizontalScrollBarValue()
+	{
+		HorizontalScrollValue = -1 * HorizontalOffset;
+		await RaiseRenderedContentUpdated();
+	}
+
+	private void IsActiveChanged(DependencyObject sender, DependencyProperty dp)
+	{
+		if (ResetWhenNotActive && !IsActive)
+		{
+			RemoveOffset();
+			ResetZoom();
+		}
+		if (_scrollH is not null)
+		{
+			_scrollH.Visibility = IsActive ? Visibility.Visible : Visibility.Collapsed;
+		}
+		if (_scrollV is not null)
+		{
+			_scrollV.Visibility = IsActive ? Visibility.Visible : Visibility.Collapsed;
+		}
+	}
+
+	private void UpdateScrollLimits()
+	{
+		if (_presenter?.Content is FrameworkElement fe)
+		{
+			// If Actual Width/Height * Zoom < ViewPort Width/Height there should be no scroll
+			// If Actual Width/Height * Zoom > ViewPort Width/Height the amount of scroll should
+			// be based on the difference between the two values
+			var verticalScroll = Math.Max(0, (fe.ActualHeight * ZoomLevel) - ViewPortHeight);
+			var horizontalScroll = Math.Max(0, (fe.ActualWidth * ZoomLevel) - ViewPortWidth);
+
+			HorizontalMaxScroll = horizontalScroll / 2;
+			VerticalMaxScroll = verticalScroll / 2;
+
+			HorizontalMinScroll = -1 * HorizontalMaxScroll;
+			VerticalMinScroll = -1 * VerticalMaxScroll;
+		}
+	}
+
+	private void CoerceZoomLevel(DependencyObject sender, DependencyProperty dp)
+	{
+		var control = (ZoomContentControl)sender;
+		control.ZoomLevel = Math.Clamp(control.ZoomLevel, control.MinZoomLevel, control.MaxZoomLevel);
+	}
+
+	public ZoomContentControl()
+	{
+		DefaultStyleKey = typeof(ZoomContentControl);
+
+		RegisterPropertyHandlers();
+
+		this.SizeChanged += ZoomContentControl_SizeChanged;
+	}
+
+	private void ZoomContentControl_SizeChanged(object sender, SizeChangedEventArgs args)
+	{
+		if (AutoZoomToCanvasOnSizeChanged)
+		{
+			ZoomToCanvas();
+		}
+	}
+
+	protected override void OnApplyTemplate()
+	{
+		_grid = GetTemplateChild(Parts.RootGrid) as Grid;
+		_presenter = GetTemplateChild(Parts.Presenter) as ContentPresenter;
+		_scrollV = GetTemplateChild(Parts.VerticalScrollBar) as ScrollBar;
+		_scrollH = GetTemplateChild(Parts.HorizontalScrollBar) as ScrollBar;
+
+		if (_grid is null ||
+			_presenter is null ||
+			_scrollH is null ||
+			_scrollV is null)
+		{
+			throw new InvalidOperationException("ZoomContentControl template is missing required parts.");
+		}
```

Comment:
better to be explicit on which template-part failed to resolve:
```suggestion
		T FindTemplatePart<T>(string name) where T : class =>
			(GetTemplateChild(name) ?? throw new Exception($"Expected template part not found: {name}"))
			as T ?? throw new Exception($"Expected template part '{name}' to be of type: {typeof(T)}");

		_grid = FindTemplatePart<Grid>(Parts.RootGrid);
		_presenter = FindTemplatePart<ContentPresenter>(Parts.Presenter);
		_scrollV = FindTemplatePart<ScrollBar>(Parts.VerticalScrollBar);
		_scrollH = FindTemplatePart<ScrollBar>(Parts.HorizontalScrollBar);
```